docs_dev/kubeflow_deployment.md

```diff
@@ -0,0 +1,282 @@
+for multiple components going forward
+
+   * A user should be able to manage the components individually; e.g. updating JupyterHub without updating other components
+   * We end up combining the parameters for all the components into a single prototype file which is a bit unwieldy
+   * It limits our ability to organize components into ksonnet packages by application (e.g. TensorFlow, PyTorch, Jupyter)
+
+## Monitoring Kubeflow Deployment
+
+Monitoring deployment is an open area. The initial focus is:
+
+  1. Informing the user when Kubeflow is fully deployed and ready to use
+  1. Surfacing relevant events (e.g. as K8s events)
+
+For more info please take a look at some of the open issues and consider
+opening more. 
+
+The current thinking is to follow the guidance of sig-apps and use an [application resource](https://github.com/kubernetes-sigs/application)
+to represent Kubeflow and attach events, status, and other metrics to that application as appropriate.
+
+
+* [kubeflow/kubeflow#1106](https://github.com/kubeflow/kubeflow/issues/1106) Use Application CRD to describe Kubeflow
+* [kubeflow/kubeflow#1142](https://github.com/kubeflow/kubeflow/issues/1142) - Report K8s events to indicate when Kubeflow is ready
+* [kubeflow/kubeflow#955](https://github.com/kubeflow/kubeflow/issues/955) - Determine when ingress to Kubeflow is ready
+
+## Testing
+
+Continuous testing of deploy scripts is critical. When our test infrastructure
+is sufficiently mature CI will likely be a requirement in order to be an officially
+recommended solution on kubeflow.org.
+
+For more information on testing
+
+* Refer to [docs](https://github.com/kubeflow/testing) for our test infrastructure
+
+# References
+
+* [kubeflow/kubeflow#105](https://github.com/kubeflow/kubeflow/issues/105) - Make it
+  easy to get started with Kubeflow
+* [kubeflow/kubeflow#23](https://github.com/kubeflow/kubeflow/issues/23) - Original issue about tooling and configuration.
+* [kubeflow/kubeflow#376](https://github.com/kubeflow/kubeflow/issues/376) - Discussion of ksonnet patterns
+
+# Appendix: Other Approaches to Customization
+
+## Auto Configuration
+
+One of the ideas we considered was creating a simple program (originally called bootstrapper) that could automatically optimize a 
+Kubeflow configuration
+based on a user's setup. For example, if the K8s cluster has a default storage
+class the program would automatically configure JupyterHub to use persistent 
+volumes for notebook storage.
+
+So far this idea hasn't gained much traction. 
+
+Most of the setup to date has focused on users starting from scratch. In this situation we 
+have complete control of the setup so there's no reason to optimize with respect to an
+existing deployment.
+
+The scaffolding/prototype for this (bootstrapper) is still in place and we haven't
+rejected this idea completely. So contributions pursuing this idea further would
+be welcome.
```

Comment:
@jlewi I liked the design of the bootstrapper running in the 'kubeflow' namespace with elevated privileges. It did seem to have a limitation of only creating/deploying one kubeflow instance/namespace with no specific rbac mapping to users in that namespace. i wonder if this design might be continued with a local command communicating with the bootstrapper for create, delete, list, add <user/group>, and other subcommands. The bootstrapper would leverage the `application` as a bundling construct for different types of applications that have different users, cloud-providers and components.